staging/src/k8s.io/apiserver/pkg/server/routes/openstack.go

```diff
@@ -30,30 +31,100 @@ import (
 const (
 	POD_URL_TEMPLATE       = "/api/v1/tenants/%s/namespaces/%s/pods"
 	OPENSTACK_SERVERS_PATH = "/servers"
+	OPENSTACK_FLAVORS_PATH = "/flavors"
+	OPENSTACK_IMAGES_PATH  = "/images"
+
+	TARGET_FLAVORS = "flavors"
+	TARGET_IMAGES  = "images"
 )
 
 type Openstack struct{}
 
 // the url path is /servers/{vmId}
-func getVmFromPath(path string) string {
+func getElementFromPath(path string) string {
 	return strings.Split(path, "/")[2]
 }
 
+func (o Openstack) imageHandler(resp http.ResponseWriter, req *http.Request) {
+	o.genericFunc(TARGET_IMAGES, resp, req)
+}
+
+func (o Openstack) flavorHandler(resp http.ResponseWriter, req *http.Request) {
+	o.genericFunc(TARGET_FLAVORS, resp, req)
+}
+
+func (o Openstack) genericFunc(target string, resp http.ResponseWriter, req *http.Request) {
+	path := req.URL.Path
+	klog.V(4).Infof("URL path: %s", path)
+
+	if req.Method != "GET" {
+		resp.WriteHeader(405) // method not allowed for the current release
+		return
+	}
+
+	var body []byte
+	var err error
+	var f interface{}
+
+	getter := func(name string) (interface{}, error) {
+		switch target {
+		case TARGET_FLAVORS:
+			return openstack.GetFalvor(name)
+		case TARGET_IMAGES:
+			return openstack.GetImage(name)
+		default:
+			return nil, fmt.Errorf("invalid target %s", target)
+		}
+	}
+
+	lister := func() interface{} {
+		switch target {
+		case TARGET_FLAVORS:
+			return openstack.ListFalvors()
+		case TARGET_IMAGES:
+			return openstack.ListImages()
+		default:
+			return nil
+		}
+	}
+
+	if strings.HasSuffix(path, target) {
+		body, err = json.Marshal(lister())
+	} else {
+		name := getElementFromPath(path)
+		f, err = getter(name)
+		if err != nil {
+			klog.Errorf("Invalid %s %s", target, name)
+			resp.WriteHeader(500)
+			return
+		}
+		body, err = json.Marshal(f)
+	}
+
+	if err != nil {
+		klog.Errorf("failed encoding %s, error %v", target, err)
+		resp.WriteHeader(500)
```

Comment:
Can we use http.StatusInternalServerError instead of 500? 405 as well

Reply:
good point. will change.

Reply:
fixed.